Api 33483 poa v2 signatures

acovrig opened the pull request:
## Summary
- Require rep registrationNumber

## Related issue(s)
- [API-33483](https://jira.devops.va.gov/browse/API-33483)

## Testing done
- [x] *New code is covered by unit tests*
- Postman
  - POST v2 POA validate

## What areas of the site does it impact?
v2 POA

## Acceptance criteria
- [x]  I fixed|updated|added unit tests and integration tests for each feature (if applicable).
- [ ]  No error nor warning in the console.
- [ ]  Events are being sent to the appropriate logging solution
- [x]  Documentation has been updated (link to documentation)
- [ ]  No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs
- [ ]  Feature/bug has a monitor built into Datadog or Grafana (if applicable)
- [ ]  If app impacted requires authentication, did you login to a local build and verify all authenticated routes work as expected
- [ ]  I added a screenshot of the developed feature

# Things to know about this PR
This adds representative_id checking, meaning the test body us devs use may not work. I recommend running this from our spec against your local dev to get a working set of IDs you can use in local testing:
```
Veteran::Service::Representative.create!(representative_id: '67890', poa_codes: ['083'], first_name: 'George', last_name: 'Washington')
Veteran::Service::Organization.create!(poa: '083', name: "083 - DISABLED AMERICAN VETERANS")
```
Then you should be able to hit local with a `poa_code` of `083` and `registrationNumber` of `67890`.

acovrig commented on modules/claims_api/spec/requests/v2/veterans/rswag_power_of_attorney_spec.rb:
```diff
@@ -632,37 +639,112 @@
+        end
+      end
+
+      describe 'Getting a 404 response' do
```
Should this be a 404?

FonzMP replied:
I would have to look at what exactly is happening once we get the /validate docs merged in. it does look like this file has been changed a good bit, so we may have this resolved?

FonzMP commented on modules/claims_api/spec/requests/v2/veterans/power_of_attorney_request_spec.rb:
```diff
@@ -144,6 +144,57 @@
+            },
+            signatures: {
+              veteran: b64_image,
+              representative: b64_image
```
my question as well. My understanding is we pull everything internal our system, so we should no longer require nor need the first and last name, or the signatures provided. Was curious as well when I saw the changes to 2122/valid.json above. 

```
"signatures": {
    "veteran": "",
    "representative": ""
}
```


```
- If valid registration number, include Representative's electronic signature on PDF based on first and last name found in our DB
- Include the Veteran's electronic signature on PDF based on first and last name found in MPI
```

FonzMP commented on modules/claims_api/app/swagger/claims_api/v1/swagger.json:
```diff
@@ -9476,7 +9476,7 @@
                     {
                       "title": "Unprocessable Entity",
-                      "detail": "Unable to locate Veteran's File Number in Master Person Index (MPI). Please submit an issue at ask.va.gov or call 1-800-MyVA411 (800-698-2411) for assistance.",
+                      "detail": "The SSN provided does not match Master Person Index (MPI). Please correct the SSN or submit an issue at ask.va.gov or call 1-800-MyVA411 (800-698-2411) for assistance.",
```
we can ignore this for this PR to cut down on LoC. Although it's not much. I also think this is in another PR so it can be resolved there.

FonzMP commented on modules/claims_api/app/swagger/claims_api/v2/dev/swagger.json:
```diff
@@ -10060,14 +10061,524 @@
         "responses": {
           "200": {
-            "description": "Valid request response"
+            "description": "Valid request response",
+            "content": {
+              "application/json": {
+                "example": {
+                  "data": {
+                    "type": "form/21-22/validation",
+                    "attributes": {
+                      "status": "valid"
+                    }
+                  }
+                },
+                "schema": {
+                  "type": "object",
+                  "properties": [
+                    {
+                      "status": {
+                        "type": "string"
+                      }
+                    }
+                  ]
+                }
+              }
+            }
```
I think the validate docs marged this morning. May want to update this branch. @acovrig

FonzMP commented on modules/claims_api/app/controllers/claims_api/v2/veterans/power_of_attorney_controller.rb:
```diff
@@ -26,7 +26,8 @@ def show
           shared_form_validation('2122')
-          poa_code = parse_and_validate_poa_code('2122')
+          validate_registration_number!
+          poa_code = parse_and_validate_poa_code(FORM_NUMBER)
```
we use the constant here but not a few lines above. Wondering if we should align these and maybe create one for `INDIVIDUAL` and `ORGANIZATION` (or Rep?) to make more sense? I think we may be dependent on `FORM_NUMBER` elsewhere, but should we align these values throughout the controller where possible? 

thoughts on this as well @dellerbie ?

FonzMP replied:
could tack the above changes on to a ticket specific to cleaning these up if we so desire.

FonzMP commented on modules/claims_api/app/controllers/claims_api/v2/veterans/power_of_attorney_controller.rb:
```diff
@@ -45,17 +46,18 @@ def submit2122a
+          shared_form_validation('2122')
 
-          poa_code = form_attributes.dig('serviceOrganization', 'poaCode')
+          poa_code = get_poa_code('2122A')
```
is this supposed to be 2122A? it's in a `validate2122` method and the shared validation gets `2122`?

FonzMP commented on modules/claims_api/app/controllers/claims_api/v2/veterans/power_of_attorney_controller.rb:
```diff
@@ -95,28 +97,35 @@ def submit_power_of_attorney(poa_code, form_number)
+            )
+          end
+
+          if (individuals&.pluck(:representative_id)&.uniq || []).count > 1
```
do we need the ` || [] ` here? doesn't `individuals` return empty and the line above raises an error?

even if it makes it to this line, we would get an empty array and a count of 0.
```
3.2.2 :003 > [].pluck(:representative_id).uniq
 => []
3.2.2 :004 > [].pluck(:representative_id).uniq.count
 => 0
```

jvcAdHoc commented on modules/claims_api/app/controllers/claims_api/v2/veterans/power_of_attorney_controller.rb:
```diff
@@ -89,28 +97,35 @@ def submit_power_of_attorney(poa_code, form_number)
+            )
+          end
+
+          if individuals&.pluck(:representative_id)&.uniq&.count&.> 1
```
passing the `.empty?` check should mean the nil safe operators aren't needed

jvcAdHoc commented on modules/claims_api/spec/fixtures/v2/veterans/power_of_attorney/2122/valid.json:
```diff
@@ -2,7 +2,12 @@
+        "poaCode": "083",
+        "registrationNumber": "67890"
+      },
+      "signatures": {
```
I think an older comment asked about this. I thought we were removing the signatures param entirely for POA V2.

jvcAdHoc commented on modules/claims_api/config/schemas/v2/2122.json:
the signatures attribute on line 350 should be removed

jvcAdHoc commented on modules/claims_api/spec/fixtures/v2/veterans/power_of_attorney/2122/invalid_poa.json:
```diff
@@ -2,7 +2,12 @@
+        "poaCode": "aaa",
+        "registrationNumber": "67890"
+      },
+      "signatures": {
```
You won't need signatures once it's removed from the schema

FonzMP commented on modules/claims_api/app/swagger/claims_api/v2/dev/swagger.json:
```diff
@@ -9537,7 +9476,7 @@
                       "title": "Resource not found",
                       "status": "404",
-                      "detail": "Could not find an Organization with code: 083",
+                      "detail": "Could not retrieve Power of Attorney with that registration number: invalid",
```
this is interesting. two notes
1. "with that registration number" I think should be adjusted to "with registration number"
2. Invalid doesn't look good here. we could use one from the mocking like 083

FonzMP commented:
is this waiting on another branch to merge before these changes go in? Or are we just waiting for the comments above to be addressed and resolved @acovrig ?

acovrig commented:
> is this waiting on another branch to merge before these changes go in? Or are we just waiting for the comments above to be addressed and resolved @acovrig ?

Finishing testing and addressing those comments

jvcAdHoc commented on modules/claims_api/app/sidekiq/claims_api/v1/poa_form_builder_job.rb:
```diff
@@ -47,15 +48,35 @@ def pdf_constructor(poa_code)
+        res
+      end
+
+      def text_signatures(power_of_attorney, rep)
```
I think these changes should be in the org builder class. The signatures for 2122 and 2122a are not all in the same location.

jvcAdHoc replied:
Or there should be logic to pick the org related signature attributes and or 2122a signature attributes

jvcAdHoc replied:
Yes, now that the builder is version there shouldn't be any changes in this file

FonzMP commented on modules/claims_api/app/controllers/claims_api/v2/veterans/power_of_attorney/organization_controller.rb:
```diff
@@ -15,14 +15,16 @@ def submit
           shared_form_validation(FORM_NUMBER)
           poa_code = get_poa_code(FORM_NUMBER)
           validate_org_poa_code!(poa_code)
+          validate_registration_number!
```
can we not put this in the shared validation and return the rep at :15, for use in :20?

jvcAdHoc commented on modules/claims_api/app/controllers/claims_api/v2/veterans/power_of_attorney/base_controller.rb:
```diff
@@ -42,7 +42,21 @@ def shared_form_validation(form_number)
 
-        def submit_power_of_attorney(poa_code, form_number)
+        def validate_registration_number!
+          rn = form_attributes.dig('serviceOrganization', 'registrationNumber')
```
I'm assuming the 2122a signatures ticket will update this to work for the `representative` object?

jvcAdHoc commented on modules/claims_api/config/schemas/v2/2122.json:
do we still need first and last name?

jvcAdHoc commented on modules/claims_api/app/controllers/claims_api/v2/veterans/power_of_attorney/base_controller.rb:
```diff
@@ -44,6 +44,24 @@ def request_representative
+          poa_code = form_attributes.dig(base, 'poaCode')
+          rep = ::Veteran::Service::Representative.where('? = ANY(poa_codes) AND representative_id = ?',
+                                                         poa_code,
+                                                         rn).first
```
I think `.first` will default to ordering by the `representative_id`. Our default has been to order by `created_at: :desc`.

jvcAdHoc commented on modules/claims_api/app/sidekiq/claims_api/v2/poa_form_builder_job.rb:
```diff
@@ -15,10 +15,11 @@ class PoaFormBuilderJob < ClaimsApi::ServiceBase
-      def perform(power_of_attorney_id, form_number)
+      def perform(power_of_attorney_id, form_number, rep_id = nil)
         power_of_attorney = ClaimsApi::PowerOfAttorney.find(power_of_attorney_id)
+        rep = ::Veteran::Service::Representative.find(rep_id) unless rep_id.nil?
```
Similar thing here, this might not fetch the "right" representative record

acovrig replied:
That should be the rep's primary_key GUID, so it should be the correct record.

jvcAdHoc replied:
the primary key for reps is not unique

FonzMP replied:
so we should move to something like the controller, correct? using the find_by for all and then checking last updated? using _that_ value?

jvcAdHoc replied:
Yeah, ordering a where clause by `created_at: :desc` should ensure it's the same record getting accessed. This should be an edge case. There probably aren't that many rep records with the same rep_id. However, I think for these signatures it matters to get the correct first and last name of the rep.

jvcAdHoc replied:
There are 117 rep_ids in staging with more than 1 record

jvcAdHoc commented on modules/claims_api/app/sidekiq/claims_api/v2/poa_form_builder_job.rb:
```diff
@@ -15,10 +15,13 @@ class PoaFormBuilderJob < ClaimsApi::ServiceBase
+      def perform(power_of_attorney_id, form_number, rep_id = nil)
         power_of_attorney = ClaimsApi::PowerOfAttorney.find(power_of_attorney_id)
+        unless rep_id.nil?
+          rep = ::Veteran::Service::Representative.where(representative_id: rep_id).order(updated_at: :desc).first
```
```suggestion
          rep = ::Veteran::Service::Representative.where(representative_id: rep_id).order(created_at: :desc).first
```

jvcAdHoc commented on modules/claims_api/app/sidekiq/claims_api/v2/poa_form_builder_job.rb:
```diff
@@ -76,7 +77,7 @@ def organization_signatures(power_of_attorney)
             },
             {
-              'signature' => "#{first_name} #{last_name} - signed via api.va.gov",
+              'signature' => "#{rep.first_name} #{rep.last_name} - signed via api.va.gov",
```
We know in V2 that the incoming `rep_id` should not be nil at this point which means we should always have the `rep` object too. I think having `rep_id = nil` in the job signature and the `unless rep_id.nil?` check could cause a confusing interpretation of not needing to pass in the rep_id. I believe the code would break here in that case as rep wouldn't be defined.

acovrig replied:
Good catch, I was originally making it an optional parameter but I guess I changed part of it when it became a required param

acovrig commented on modules/claims_api/app/controllers/claims_api/v2/veterans/power_of_attorney/base_controller.rb:
```diff
@@ -116,6 +135,11 @@ def representative(poa_code)
           format_representative(individuals.first)
         end
 
+        def organization
```
It was used in the representative method, but a merge changed that at some point... I'll remove this then.

jvcAdHoc replied:
This might have been missed in the merge conflicts

jvcAdHoc commented on modules/claims_api/config/schemas/v2/2122.json:
```diff
@@ -8,7 +8,7 @@
   ],
   "properties": {
-    "veteran": {
+   "veteran": {
```
Was this indentation change intentional?

jvcAdHoc commented on modules/claims_api/config/schemas/v2/2122.json:
```diff
@@ -218,15 +219,10 @@
           "type": "string",
           "example": "I help vets LLC."
         },
-        "firstName": {
```
I realized with removing first and last name that the form constructor needs to change. `modules/claims_api/lib/claims_api/v2/poa_pdf_constructor/organization.rb` item 16A uses the first and last name attributes
